File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/composer/TelexState.kt
package com.draftright.keyboard.composer

object TelexState {
    private val PLAIN_VOWELS = setOf(
        'a', 'e', 'i', 'o', 'u', 'y',
        'A', 'E', 'I', 'O', 'U', 'Y',
    )
    private val SPECIAL_VOWELS = setOf(
        'ă', 'â', 'ê', 'ô', 'ơ', 'ư',
        'Ă', 'Â', 'Ê', 'Ô', 'Ơ', 'Ư',
    )
    private val TONE_MARKS = setOf('s', 'f', 'r', 'x', 'j')

    /**
     * Vowels that can sit AFTER the nucleus inside one cluster without being a
     * nucleus themselves. Lives here with the other vowel classifications so
     * there is one place that answers "what kind of vowel is this" (#152).
     */
    private val GLIDE_CODAS = setOf('i', 'y', 'u', 'o', 'I', 'Y', 'U', 'O')

    fun isVowel(c: Char): Boolean = c in PLAIN_VOWELS
    fun isVowelLike(c: Char): Boolean = c in PLAIN_VOWELS || c in SPECIAL_VOWELS
    fun isSpecialVowel(c: Char): Boolean = c in SPECIAL_VOWELS
    fun isToneMark(c: Char): Boolean = c in TONE_MARKS

    /**
     * True for an offglide — a vowel a trailing modifier may reach back over
     * to find its nucleus ("day" + a: the 'y' is an offglide, the 'a' is the
     * nucleus and takes the circumflex). Reaching past a non-glide vowel would
     * corrupt genuine clusters like "oeo" (ngoẻo).
     */
    fun isGlideCoda(c: Char): Boolean = c in GLIDE_CODAS
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Composer/TelexState.swift
import Foundation

public enum TelexState {
    static let plainVowels: Set<Character> = [
        "a", "e", "i", "o", "u", "y",
        "A", "E", "I", "O", "U", "Y",
    ]
    static let specialVowels: Set<Character> = [
        "ă", "â", "ê", "ô", "ơ", "ư",
        "Ă", "Â", "Ê", "Ô", "Ơ", "Ư",
    ]
    /// Vowels that can sit AFTER the nucleus as an offglide (semivowel coda) —
    /// the trailing i/y/u/o of "day", "tai", "keo", "cau". A modifier typed at
    /// the end may reach back over these to the nucleus, but not over another
    /// full nucleus vowel. Mirrors Android's TelexState.GLIDE_CODAS.
    static let glideCodas: Set<Character> = [
        "i", "y", "u", "o",
        "I", "Y", "U", "O",
    ]
    static let toneMarks: Set<Character> = ["s", "f", "r", "x", "j"]

    public static func isVowel(_ c: Character) -> Bool {
        plainVowels.contains(c)
    }

    public static func isVowelLike(_ c: Character) -> Bool {
        plainVowels.contains(c) || specialVowels.contains(c)
    }

    public static func isSpecialVowel(_ c: Character) -> Bool {
        specialVowels.contains(c)
    }

    public static func isGlideCoda(_ c: Character) -> Bool {
        glideCodas.contains(c)
    }

    public static func isToneMark(_ c: Character) -> Bool {
        toneMarks.contains(c)
    }
}
